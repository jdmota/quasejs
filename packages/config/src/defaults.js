// @flow
import Path from "./path";
import { t, toType } from "./types";

export function clone( path: Path, _type: any, cloned: any, src: any, key: any ) {

  path.push( key );

  const type = toType( _type, path );
  const { map } = type.extra;
  const srcValue = map ? map( src[ key ] ) : src[ key ];

  if ( srcValue === undefined ) {
    return;
  }

  const newValue = type.clone( path, srcValue );

  path.pop();

  if ( newValue !== undefined ) {
    cloned[ key ] = newValue;
  }
}

export function merge( path: Path, _type: any, object: any, src: any, key: any ) {

  path.push( key );

  const type = toType( _type, path );
  const { map, merge } = type.extra;
  const objValue = object[ key ];
  const srcValue = map ? map( src[ key ] ) : src[ key ];

  if ( srcValue === undefined ) {
    return;
  }

  if ( objValue !== undefined && merge === "override" ) {
    return;
  }

  let newValue;

  if ( objValue === undefined ) {
    newValue = type.clone( path, srcValue );
  } else {
    newValue = type.merge( path, objValue, srcValue );
  }

  path.pop();

  if ( newValue !== undefined ) {
    object[ key ] = newValue;
  }
}

function extractDefaultsHelper( path: Path, _type: any, defaults: any, key: any, dest: Object ) {
  const type = toType( _type, path );
  const { default: d, required, optional } = type || {};
  if ( required && optional ) {
    throw new Error( `[Schema] Don't use "required" and "optional" in ${path.chainToString()}` );
  }
  if ( d === undefined ) {
    if ( required || optional ) {
      return;
    }
  } else {
    if ( required ) {
      throw new Error( `[Schema] Don't use "required" with "default" in ${path.chainToString()}` );
    }
    return d;
  }
  const customDefault = type.defaults( path, dest );
  if ( customDefault !== undefined ) {
    return customDefault;
  }
  if ( !optional ) {
    throw new Error( `[Schema] Provide a default value or mark as "optional" in ${path.chainToString()}` );
  }
}

export function extractDefaults( path: Path, _type: any, defaults: any, key: any, dest: Object ) {

  path.push( key );

  const d = extractDefaultsHelper( path, _type, defaults, key, dest );

  if ( d !== undefined ) {
    defaults[ key ] = d;
  }

  path.pop();
}

export function applyDefaults( schema: Object, args: Object[], where: string[] = [] ) {
  const path = new Path();
  const type = t.object( { properties: schema } );
  if ( args.length > 0 ) {
    path.where = where[ 0 ];
    const dest = type.clone( path, args[ 0 ] );
    for ( let i = 1; i < args.length; i++ ) {
      path.where = where[ i ];
      type.merge( path, dest, args[ i ] );
    }
    path.where = null;
    type.merge( path, dest, type.defaults( path, dest ) );
    type.validate( path, dest, dest );
    return dest;
  }
  return type.defaults( path, {} );
}
