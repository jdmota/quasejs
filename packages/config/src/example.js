import { validate, printError, t } from ".";

try {
  validate( {
    deprecated: t.object( {
      properties: {
        deprecated: true
      }
    } )
  }, {
    deprecated: true
  } );
} catch ( e ) {
  printError( e );
}

try {
  validate( {
    config: {
      type: "string"
    }
  }, {
    config: true
  } );
} catch ( e ) {
  printError( e );
}

try {
  validate( {
    obj: t.object( {
      properties: {
        foo: t.object( {
          properties: {
            bar: {
              type: "string",
              example: "example"
            }
          }
        } )
      }
    } )
  }, {
    obj: {
      foo: {
        bar: 10
      }
    }
  } );
} catch ( e ) {
  printError( e );
}

try {
  validate( {
    obj: t.object( {
      properties: {
        foo: t.object( {
          properties: {
            bar: {
              type: "string",
              example: "example"
            }
          }
        } )
      }
    } )
  }, {
    obj: {
      foo: 10
    }
  } );
} catch ( e ) {
  printError( e );
}

try {
  validate( {
    obj: t.object( {
      properties: {
        foo: t.object( {
          properties: {
            bar: {
              type: "string",
              example: "example"
            }
          }
        } )
      }
    } )
  }, {
    obj: {
      foo: {
        baz: "abc"
      }
    }
  } );
} catch ( e ) {
  printError( e );
}

try {
  validate( {
    obj: t.object( {
      properties: {
        foo: t.tuple( {
          items: [
            {
              type: "string",
              example: "example"
            }, {
              type: "string",
              example: "example"
            }
          ]
        } )
      }
    } )
  }, {
    obj: {
      foo: [ "string", 10 ]
    }
  } );
} catch ( e ) {
  printError( e );
}

try {
  validate( {
    obj: t.object( {
      properties: {
        foo: t.tuple( {
          items: [
            {
              type: "string",
              example: "example"
            }, {
              type: "string",
              example: "example"
            }
          ]
        } )
      }
    } )
  }, {
    obj: {
      foo: {}
    }
  } );
} catch ( e ) {
  printError( e );
}

try {
  validate( {
    obj: t.object( {
      properties: {
        foo: t.union( {
          types: [
            "string",
            "number"
          ],
          example: "example"
        } )
      }
    } )
  }, {
    obj: {
      foo: [ "string", 10 ]
    }
  } );
} catch ( e ) {
  printError( e );
}

try {
  validate( {
    obj: t.object( {
      properties: {
        foo: t.choices( {
          values: [
            0,
            1
          ],
          example: 1
        } )
      }
    } )
  }, {
    obj: {
      foo: 2
    }
  } );
} catch ( e ) {
  printError( e );
}

// yarn n packages/config/src/example
