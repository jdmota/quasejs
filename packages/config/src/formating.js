// @flow
import type Path from "./path";
import { type GeneralType, types } from "./types";

const concordance = require( "concordance" );
const turbocolor = require( "turbocolor" );

const reIndent = /^(?!\s*$)/mg;

export function indent( str: string, str2: string = "  " ) {
  return str.replace( reIndent, str2 );
}

export function addPrefix( path: Path, key: string ) {
  return path.size() ? `${path.chainToString()}.${key}` : key;
}

export function formatTypes( list: $ReadOnlyArray<?GeneralType>, separator: string = " | " ): string {
  return list.filter( Boolean ).map( x => {
    if ( typeof x === "string" ) {
      return x;
    }
    if ( x != null ) {
      if ( x instanceof types.Type ) {
        if ( x instanceof types.Union ) {
          return formatTypes( x.types );
        }
        if ( x instanceof types.Value ) {
          return format( x.value );
        }
        if ( x instanceof types.Tuple ) {
          return formatTypes( x.items, ", " );
        }
        if ( x instanceof types.Array ) {
          return "array";
        }
        if ( x instanceof types.Object ) {
          return "object";
        }
      } else if ( typeof x.type === "string" ) {
        return x.type;
      }
    }
    return "";
  } ).join( separator );
}

export function formatOption( option: string ): string {
  return turbocolor.bold( concordance.format( option ) );
}

export function format( value: any ) {
  return concordance.format( value );
}
