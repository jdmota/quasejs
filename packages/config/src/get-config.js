const findUp = require( "find-up" );
const pkgConf = require( "pkg-conf" );

export async function getConfig( opts ) {

  const { cwd, configFiles, configKey, failIfNotFound } = opts || {};
  const result = {
    config: undefined,
    location: undefined
  };

  if ( configFiles && configFiles.length ) {
    const location = await findUp( configFiles, { cwd } );

    if ( location ) {
      try {
        const e = require( location );
        result.config = e && e.__esModule ? e.default : e;
        result.location = location;
      } catch ( e ) {
        // Ignore
      }
    } else if ( failIfNotFound ) {
      throw new Error( `Config file was not found: ${configFiles.toString()}` );
    }
  }

  if ( !result.config ) {
    if ( configKey ) {
      try {
        result.config = await pkgConf( configKey, { cwd, skipOnFalse: true } );
        result.location = "pkg";
      } catch ( e ) {
        // Ignore
      }
    }
  }

  return result;
}
