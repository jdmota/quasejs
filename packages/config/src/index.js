// @flow
import Path from "./path";
import getType from "./get-type";
import { t, types, toType, type Type, type Schema } from "./types";
import { ValidationError, checkType } from "./validation";

export function validate( schema: Schema, config: ?Object ) {
  config = config || {};
  t.object( { properties: schema } ).validate( new Path(), config, config );
}

export { applyDefaults } from "./defaults";
export { getConfig } from "./get-config";
export { printWarning, printError } from "./print";

export { t, types, toType, getType, ValidationError, checkType };

export function extractDefaults( type: Type, dest: ?Object ) {
  return type.defaults( new Path(), dest || {} );
}
