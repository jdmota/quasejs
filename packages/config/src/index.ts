const findUp = require( "find-up" );
const pkgConf = require( "pkg-conf" );

export function loadConfigFrom( location: string, arg: unknown ): unknown {
  const e = require( location );
  const x = e && e.__esModule ? e.default : e;
  return typeof x === "function" ? x( arg ) : x;
}

type Options = {
  cwd?: string;
  configFiles?: string | string[];
  arg?: unknown;
  configKey?: string;
  failIfNotFound?: boolean;
};

export async function getConfig( opts: Options = {} ) {

  const { cwd, configFiles, arg, configKey, failIfNotFound } = opts;
  let config: unknown;
  let location: string | undefined;

  if ( configFiles && configFiles.length ) {
    const loc = await findUp( configFiles, { cwd } ) as string;

    if ( loc ) {
      config = await loadConfigFrom( loc, arg );
      location = loc;
    } else if ( failIfNotFound ) {
      throw new Error( `Config file was not found: ${configFiles.toString()}` );
    }
  }

  if ( !config && configKey ) {
    config = await pkgConf( configKey, { cwd, skipOnFalse: true } );
    location = pkgConf.filepath( config ) as string;

    if ( location == null ) {
      config = undefined;
      location = undefined;
    }
  }

  return {
    config,
    location
  };
}
