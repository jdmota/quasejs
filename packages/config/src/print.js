// @flow
/* eslint-disable no-console */

const chalk = require( "chalk" );

export function printWarning( str: string ) {
  console.warn( `${chalk.yellow( str )}\n` );
}

export function printError( error: Error ) {
  let message;
  if ( error.__validation ) {
    message = `${chalk.bold( "Validation Error" )}:\n\n${error.message.replace( /^(?!$)/mg, "  " )}`;
  } else {
    message = error.stack;
  }
  console.error( `${chalk.red( message )}\n` );
  process.exitCode = 1;
}
