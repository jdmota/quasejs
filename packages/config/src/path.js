// @flow
import { formatOption } from "./formating";

export default class Path {

  +arr: string[];
  where: ?string;

  constructor() {
    this.arr = [];
    this.where = null;
  }

  push( key: string | number ) {
    this.arr.push( key + "" );
  }

  pop() {
    this.arr.pop();
  }

  size() {
    return this.arr.length;
  }

  chainToString() {
    return this.arr.join( "." );
  }

  format() {
    return formatOption( `${this.arr.join( "." )}${this.where ? ` [in ${this.where}]` : ""}` );
  }

}
