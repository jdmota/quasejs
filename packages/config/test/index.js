import { t, validate, printError, applyDefaults, getConfig } from "../src";

const stripAnsi = require( "strip-ansi" );

/* eslint no-console: 0 */

function v( options, schema ) {
  const mock = jest.fn();
  console.error = mock;
  console.warn = mock;
  try {
    validate( schema, options );
    mock( "OK" );
  } catch ( e ) {
    printError( e );
  }
  expect(
    mock.mock.calls.map( x => {
      if ( Array.isArray( x ) ) {
        return x.map( stripAnsi );
      }
      return stripAnsi( x );
    } )
  ).toMatchSnapshot();
}

it( "validate", () => {

  v( {
    deprecated: true
  }, {
    deprecated: {
      deprecated: true
    }
  } );

  v( {
    config: true
  }, {
    config: {
      type: "string"
    }
  } );

  v( {
    obj: {
      foo: {
        bar: 10
      }
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          type: t.object( {
            bar: {
              type: "string",
              example: "example"
            }
          } )
        }
      } )
    }
  } );

  v( {
    obj: {
      foo: 10
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          type: t.object( {
            bar: {
              type: "string",
              example: "example"
            }
          } )
        }
      } )
    }
  } );

  v( {
    obj: {
      foo: {
        baz: "abc"
      }
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          type: t.object( {
            bar: {
              type: "string",
              example: "example"
            }
          } )
        }
      } )
    }
  } );

  v( {
    obj: {
      foo: [ "string", 10 ]
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          type: t.tuple( [
            {
              type: "string",
              example: "example"
            }, {
              type: "string",
              example: "example"
            }
          ] )
        }
      } )
    }
  } );

  v( {
    obj: {
      foo: {}
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          type: t.tuple( [
            {
              type: "string",
              example: "example"
            }, {
              type: "string",
              example: "example"
            }
          ] )
        }
      } )
    }
  } );

  v( {
    obj: {
      foo: [ "string", 10 ]
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          type: t.union( [
            "string",
            "number"
          ] ),
          example: "example"
        }
      } )
    }
  } );

  v( {
    obj: {
      foo: 2
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          choices: [
            0,
            1
          ],
          example: 1
        }
      } )
    }
  } );

  v( {
    required: undefined
  }, {
    required: {
      type: "boolean",
      required: true
    }
  } );

  v( {
    optional: undefined
  }, {
    optional: {
      type: "boolean",
      optional: true
    }
  } );

  v( {
    required: undefined
  }, {
    required: {
      type: "boolean"
    }
  } );

  v( {
    obj: {
      foo: 2
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          type: t.union(
            [ 0, 1, "object" ].map( t.value )
          ),
          example: 1
        }
      } )
    }
  } );

  v( {
    obj: {
      foo: 2
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          type: t.union(
            [ t.value( 0 ), t.value( 1 ), "object" ]
          ),
          example: 1
        }
      } )
    }
  } );

  v( {
    obj: {
      foo: 0
    }
  }, {
    obj: {
      type: t.object( {
        foo: {
          type: t.union(
            [ t.value( 0 ), t.value( 1 ), "object" ]
          ),
          example: 1
        }
      } )
    }
  } );

} );

function d( schema, ...args ) {
  try {
    expect( applyDefaults( schema, ...args ) ).toMatchSnapshot();
  } catch ( err ) {
    expect( stripAnsi( err.message ) ).toMatchSnapshot();
  }
}

it( "apply defaults", () => {

  d( {
    obj: {
      type: t.object( {
        foo: {
          type: t.tuple( [
            {
              type: "string",
              default: "a"
            }, {
              type: "string",
              default: "b"
            }
          ] )
        }
      } )
    }
  }, {
    obj: {
      foo: []
    }
  } );

  d( {
    obj: {
      type: t.object( {
        foo: {
          type: t.object( {
            a: {
              default: 10
            },
            b: {
              default: 20
            }
          } )
        }
      } )
    }
  }, {
    obj: {}
  } );

  d( {
    obj: {
      type: t.object( {
        foo: {
          type: t.object( {
            a: {
              default: 10
            },
            b: {
              default: 20
            }
          } )
        }
      } )
    }
  }, {
    obj: {
      foo: 100
    }
  } );

  d( {
    obj: {
      type: "boolean"
    }
  } );

  d( {
    obj: {
      type: "array"
    }
  } );

  d( {
    obj: {
      type: "object"
    }
  } );

  d( {
    obj: {
      type: "number",
      optional: true
    }
  } );

  d( {
    obj: {
      type: "number",
      default: 0
    }
  } );

  d( {
    obj: {
      type: "number"
    }
  } );

  d( {
    obj: {
      type: "number",
      default: 0
    }
  }, undefined, null );

} );

it( "apply defaults - merge modes", () => {

  d( {
    obj: {
      type: t.object( {
        foo: {
          type: "array",
          merge: "concat"
        }
      } )
    }
  }, {
    obj: {
      foo: [ 1 ]
    }
  }, {
    obj: {
      foo: [ 0 ]
    }
  } );

  d( {
    obj: {
      type: t.object( {
        foo: {
          type: t.array(),
          merge: "concat"
        }
      } )
    }
  }, {
    obj: {
      foo: [ 1 ]
    }
  }, {
    obj: {
      foo: [ 0 ]
    }
  } );

  d( {
    obj: {
      type: t.object( {
        foo: {
          type: t.array(),
          merge: "spreadMeansConcat"
        }
      } )
    }
  }, {
    obj: {
      foo: [ 1 ]
    }
  }, {
    obj: {
      foo: [ 0 ]
    }
  } );

  d( {
    obj: {
      type: t.object( {
        foo: {
          type: t.array(),
          merge: "spreadMeansConcat"
        }
      } )
    }
  }, {
    obj: {
      foo: [ "...", 2 ]
    }
  }, {
    obj: {
      foo: [ "...", 1 ]
    }
  }, {
    obj: {
      foo: [ 0 ]
    }
  } );

  d( {
    obj: {
      type: t.object( {
        foo: {
          type: "array",
          merge: "merge"
        }
      } )
    }
  }, {
    obj: {
      foo: [ 1 ]
    }
  }, {
    obj: {
      foo: [ 0 ]
    }
  } );

  d( {
    obj: {
      type: t.object( {
        foo: {
          type: "array",
          merge() {
            return [];
          }
        }
      } )
    }
  }, {
    obj: {
      foo: [ 1 ]
    }
  }, {
    obj: {
      foo: [ 0 ]
    }
  } );

} );

it( "get config", async() => {

  let result = await getConfig( {
    cwd: __dirname,
    configFiles: "quase-config.js",
  } );

  expect( result.config.iAmTheConfigFile ).toBe( "yes" );
  expect( typeof result.location ).toBe( "string" );

  result = await getConfig( {
    cwd: __dirname,
    configFiles: "non-existent-file.js",
  } );

  expect( result.config ).toBe( undefined );
  expect( result.location ).toBe( undefined );

  result = await getConfig( {
    cwd: __dirname,
    configFiles: "non-existent-file.js",
    configKey: "my-key"
  } );

  expect( result.config.configFromPkg ).toBe( "yes" );
  expect( result.location ).toBe( "pkg" );

  result = await getConfig( {
    cwd: __dirname,
    configFiles: "quase-config-2.js",
    configKey: "my-key"
  } );

  expect( result.config.iAmTheConfigFile2 ).toBe( "yes" );
  expect( result.config.configFromPkg ).toBe( undefined );
  expect( typeof result.location ).toBe( "string" );
  expect( result.location ).not.toBe( "pkg" );

  await expect(
    getConfig( {
      cwd: __dirname,
      configFiles: "non-existante-file.js",
      failIfNotFound: true
    } )
  ).rejects.toThrow( /^Config file was not found/ );

  result = await getConfig( {
    cwd: __dirname,
    configFiles: [],
    configKey: ""
  } );

  expect( result.config ).toBe( undefined );
  expect( result.location ).toBe( undefined );

  result = await getConfig( {
    cwd: __dirname,
    arg: 10,
    configFiles: "quase-config-3.js"
  } );

  expect( result.config.fromFunction ).toBe( "yes" );
  expect( result.config.arg ).toBe( 10 );
  expect( typeof result.location ).toBe( "string" );
  expect( result.location ).not.toBe( "pkg" );

} );
