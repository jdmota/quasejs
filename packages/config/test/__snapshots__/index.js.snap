// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply defaults - merge modes 1`] = `
Object {
  "obj": Object {
    "foo": Array [
      0,
      1,
    ],
  },
}
`;

exports[`apply defaults - merge modes 2`] = `
Object {
  "obj": Object {
    "foo": Array [
      0,
      1,
    ],
  },
}
`;

exports[`apply defaults - merge modes 3`] = `
Object {
  "obj": Object {
    "foo": Array [
      1,
    ],
  },
}
`;

exports[`apply defaults - merge modes 4`] = `
Object {
  "obj": Object {
    "foo": Array [
      0,
      1,
      2,
    ],
  },
}
`;

exports[`apply defaults - merge modes 5`] = `
Object {
  "obj": Object {
    "foo": Array [
      1,
    ],
  },
}
`;

exports[`apply defaults 1`] = `
Object {
  "obj": Object {
    "foo": Array [
      "a",
      "b",
    ],
  },
}
`;

exports[`apply defaults 2`] = `
Object {
  "obj": Object {
    "foo": Object {
      "a": 10,
      "b": 20,
    },
  },
}
`;

exports[`apply defaults 3`] = `
"Option 'obj.foo' must be of type:
  object
but instead received:
  number
"
`;

exports[`apply defaults 4`] = `
Object {
  "obj": false,
}
`;

exports[`apply defaults 5`] = `
Object {
  "obj": Array [],
}
`;

exports[`apply defaults 6`] = `
Object {
  "obj": Object {},
}
`;

exports[`apply defaults 7`] = `Object {}`;

exports[`apply defaults 8`] = `
Object {
  "obj": 0,
}
`;

exports[`apply defaults 9`] = `"[Schema] Provide a default value or mark as \\"optional\\" in obj"`;

exports[`apply defaults 10`] = `
Object {
  "obj": 0,
}
`;

exports[`apply defaults 11`] = `Object {}`;

exports[`apply defaults 12`] = `
Object {
  "obj": Object {
    "a": true,
    "b": true,
  },
}
`;

exports[`apply defaults 13`] = `
Object {
  "obj": Object {
    "a": false,
    "b": false,
  },
}
`;

exports[`apply defaults 14`] = `
Object {
  "obj": Object {
    "a": true,
    "b": true,
  },
}
`;

exports[`apply defaults 15`] = `
Object {
  "obj": Object {
    "a": false,
    "b": true,
  },
}
`;

exports[`apply defaults 16`] = `
Object {
  "minify": true,
  "mode": "production",
}
`;

exports[`show where the error is 1`] = `
"Option 'obj [in b]' must be of type:
  object
but instead received:
  array
"
`;

exports[`validate 1`] = `
Array [
  Array [
    "Deprecation Warning: Option 'deprecated' is deprecated.
",
  ],
  Array [
    "OK",
  ],
]
`;

exports[`validate 2`] = `
Array [
  Array [
    "Validation Error:

  Option 'config' must be of type:
    string
  but instead received:
    boolean

",
  ],
]
`;

exports[`validate 3`] = `
Array [
  Array [
    "Validation Error:

  Option 'obj.foo.bar' must be of type:
    string
  but instead received:
    number
  Example:
  {
    'obj': {
      'foo': {
        'bar': 'example'
      }
    }
  }
",
  ],
]
`;

exports[`validate 4`] = `
Array [
  Array [
    "Validation Error:

  Option 'obj.foo' must be of type:
    object
  but instead received:
    number

",
  ],
]
`;

exports[`validate 5`] = `
Array [
  Array [
    "Validation Error:

  Unrecognized option 'obj.foo.baz'. Did you mean 'obj.foo.bar'?
",
  ],
]
`;

exports[`validate 6`] = `
Array [
  Array [
    "Validation Error:

  Option 'obj.foo.1' must be of type:
    string
  but instead received:
    number
  Example:
  {
    'obj': {
      'foo': {
        '1': 'example'
      }
    }
  }
",
  ],
]
`;

exports[`validate 7`] = `
Array [
  Array [
    "Validation Error:

  Option 'obj.foo' must be an array of 2 items.

",
  ],
]
`;

exports[`validate 8`] = `
Array [
  Array [
    "Validation Error:

  Option 'obj.foo' should be one of these types:
    string | number
  but instead received:
    [ 'string', 10 ]
  Example:
  {
    'obj': {
      'foo': 'example'
    }
  }
",
  ],
]
`;

exports[`validate 9`] = `
Array [
  Array [
    "Validation Error:

  Option 'obj.foo' should be one of:
    0 | 1
  but instead received:
    2
",
  ],
]
`;

exports[`validate 10`] = `
Array [
  Array [
    "Validation Error:

  Option 'required' is required.
",
  ],
]
`;

exports[`validate 11`] = `
Array [
  Array [
    "OK",
  ],
]
`;

exports[`validate 12`] = `
Array [
  Array [
    "Validation Error:

  Option 'required' is required.
",
  ],
]
`;

exports[`validate 13`] = `
Array [
  Array [
    "Validation Error:

  Option 'obj.foo' should be one of these types:
    0 | 1 | 'object'
  but instead received:
    2
  Example:
  {
    'obj': {
      'foo': 1
    }
  }
",
  ],
]
`;

exports[`validate 14`] = `
Array [
  Array [
    "Validation Error:

  Option 'obj.foo' should be one of these types:
    0 | 1 | object
  but instead received:
    2
  Example:
  {
    'obj': {
      'foo': 1
    }
  }
",
  ],
]
`;

exports[`validate 15`] = `
Array [
  Array [
    "OK",
  ],
]
`;

exports[`validate 16`] = `
Array [
  Array [
    "Validation Error:

  foo and bar should not be set at the same time
",
  ],
]
`;

exports[`validate 17`] = `
Array [
  Array [
    "Validation Error:

  Unrecognized option 'foo.unknown'.
",
  ],
]
`;

exports[`validate 18`] = `
Array [
  Array [
    "OK",
  ],
]
`;

exports[`validate 19`] = `
Array [
  Array [
    "Validation Error:

  Option 'obj.foo' must be of type:
    array
  but instead received:
    object
  Example:
  {
    'obj': {
      'foo': [ 'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a',
        'a' ]
    }
  }
",
  ],
]
`;
